sentry-android-replay/src/main/java/io/sentry/android/replay/ReplayCache.kt

```diff
@@ -0,0 +1,181 @@
+        if (replayCacheDir == null) {
+            return
+        }
+
+        val screenshot = File(replayCacheDir, "$frameTimestamp.jpg").also {
+            it.createNewFile()
+        }
+        screenshot.outputStream().use {
+            bitmap.compress(JPEG, 80, it)
+            it.flush()
+        }
+
+        addFrame(screenshot, frameTimestamp)
+    }
+
+    public fun addFrame(screenshot: File, frameTimestamp: Long) {
+        val frame = ReplayFrame(screenshot, frameTimestamp)
+        frames += frame
+    }
+
+    public fun createVideoOf(
+        duration: Long,
+        from: Long,
+        segmentId: Int,
+        videoFile: File = File(replayCacheDir, "$segmentId.mp4")
+    ): GeneratedVideo? {
+        if (frames.isEmpty()) {
+            options.logger.log(
+                DEBUG,
+                "No captured frames, skipping generating a video segment"
+            )
+            return null
+        }
+
+        // TODO: reuse instance of encoder and just change file path to create a different muxer
+        encoder = synchronized(encoderLock) { encoderCreator(videoFile) }
+
+        val step = 1000 / recorderConfig.frameRate.toLong()
+        var frameCount = 0
+        var lastFrame: ReplayFrame = frames.first()
+        for (timestamp in from until (from + (duration)) step step) {
+            val iter = frames.iterator()
+            val frameCountBefore = frameCount
+            while (iter.hasNext()) {
+                val frame = iter.next()
+                if (frame.timestamp in (timestamp..timestamp + step)) {
+                    frameCount++
+                    encode(frame)
+                    lastFrame = frame
+                    break // we only support 1 frame per given interval
+                }
+            }
+
+            // if the frame count hasn't changed we just replicate the last known frame to respect
+            // the video duration.
+            if (frameCountBefore == frameCount) {
+                frameCount++
+                encode(lastFrame)
+            }
+        }
+
+        if (frameCount == 0) {
+            options.logger.log(
+                DEBUG,
+                "Generated a video with no frames, not capturing a replay segment"
+            )
+            deleteFile(videoFile)
+            return null
+        }
+
+        var videoDuration: Long
+        synchronized(encoderLock) {
+            encoder?.release()
+            videoDuration = encoder?.duration ?: 0
+            encoder = null
+        }
+
+        frames.removeAll {
+            if (it.timestamp < (from + duration)) {
+                deleteFile(it.screenshot)
+                return@removeAll true
+            }
+            return@removeAll false
+        }
+
+        return GeneratedVideo(videoFile, frameCount, videoDuration)
+    }
+
+    private fun encode(frame: ReplayFrame) {
+        val bitmap = BitmapFactory.decodeFile(frame.screenshot.absolutePath)
```

Comment:
since this is performing file IO, it probably makes sense to guard this via try/catch. e.g. for disk errors or file corruptions

Reply:
good point!